Lib/test/test_htmlparser.py

```diff
@@ -483,6 +483,18 @@ def test_comma_between_attributes(self):
                                 [('action', '/xxx.php?a=1&b=2&'),
                                  (',', None), ('method', 'post')])])
 
+    def test_comma_between_unquoted_attributes(self):
+        # bpo 41748
+        self._run_check('<div class=bar,baz=asd>',
+                        [('starttag', 'div', [('class', 'bar,baz=asd')])])
+
+    def test_comma_with_space_between_unquoted_attributes(self):
+        # bpo 41748
+        self._run_check('<div class=bar ,baz=asd>',
+                        [('starttag', 'div', [
+                            ('class', 'bar'),
+                            (',baz', 'asd')])])
+
```

Comment:
I would also add more comprehensive tests, in particular:
```
<div class=bar,baz=asd>  # between attrs (unquoted)
<div class="bar",baz="asd">  # between attrs (quoted)
<div class=bar, baz=asd,>  # after values (unquoted)
<div class="bar", baz="asd",>  # after values (quoted)
<div class=,bar baz=,asd>  # before values (unquoted)
<div class=,"bar" baz=,"asd">  # before values (quoted)
<div ,class=bar ,baz=asd>  # before names
<div class,="bar" baz,="asd">  # after names
```
Some of these are already covered by the previous tests, but they can be expanded/integrated with these suggestions.

Reply:
@ezio-melotti 

Probably these 

https://github.com/python/cpython/blob/c94ee13ad596d26d1859078bc09806aa59bb0000/Lib/test/test_htmlparser.py#L455-L489

should be moved under 

https://github.com/python/cpython/blob/c94ee13ad596d26d1859078bc09806aa59bb0000/Lib/test/test_htmlparser.py#L638-L639

That seems more logical.

Reply:
Because I had a test failing:

## case 1 - one attribute and trailing comma
input:
```
data:text/html,<!doctype html><div class="bar",>text</div>
```

browser ouput
```html
<div class="bar" ,="">text</div>
```

htmlparser output
```
 ('starttag', 'div', [('class', 'bar'), (',', None)]),
```

## case 2 - two attributes with trailing commas
Input
```
data:text/html,<!doctype html><div class="bar", baz="asd",>text</div>
```

browser  output
```html
<div class="bar" ,="" baz="asd">text</div>
```

htmlparser output - FAIL
```
 ('starttag',
  'div',
  [('class', 'bar'), (',', None), ('baz', 'asd'), (',', None)]),
```


at the beginning I thought it was a mistake, so let's try with something more common. 

## case 3 - duplicate attributes with same value:

Input
```
data:text/html,<!doctype html><div class="bar" class="bar">text</div>
```

browser output
```html
<div class="bar">text</div>
```

htmlparser output
```
 ('starttag', 'div', [('class', 'bar'), ('class', 'bar')])]
```

## case 4 - duplicate attributes with different values:

Input
```
data:text/html,<!doctype html><div class="bar" class="foo">text</div>
```

browser output
```html
<div class="bar">text</div>
```

htmlparser output
```
 ('starttag', 'div', [('class', 'bar'), ('class', 'foo')])]
```

Basically the browsers only take the first value and ignore the duplicate attribute. **but that's a different bug** and it should be solved separately.

So I will remove the test with two trailing commas. :)

Reply:
I have opened a new bug for duplicate attributes
https://bugs.python.org/issue42821

Reply:
I replied to BPO-42821 and closed it as "not a bug".
I think the output of HTMLParser is correct in all four cases and preserving the duplicated attributes is the desired behavior.
You can add back these tests, with a short comment explaining that HTMLParser preserves duplicated attributes.

Reply:
ah good point making it the role of a treebuilder and not this library. Understood.